From 5d3569a9493c7f5bb05ce15f05eebee02ae66b34 Mon Sep 17 00:00:00 2001
From: Shengjiu Wang <shengjiu.wang@nxp.com>
Date: Fri, 26 Jan 2018 16:44:08 +0800
Subject: [PATCH 3266/5242] MLK-17470: ASoC: ak4497: automatically select
 dsdsel in driver

commit  fcb6118f54cea3cb738139dc57c9f12da52783fc from
https://source.codeaurora.org/external/imx/linux-imx.git

automatically select dsdsel in driver according to the frequency

Signed-off-by: Shengjiu Wang <shengjiu.wang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 sound/soc/codecs/ak4497.c |   36 ++++++++++++++++++++++++++++++++++++
 sound/soc/codecs/ak4497.h |   13 +++++++++++++
 2 files changed, 49 insertions(+)

diff --git a/sound/soc/codecs/ak4497.c b/sound/soc/codecs/ak4497.c
index 74145a6..1eb6d40 100644
--- a/sound/soc/codecs/ak4497.c
+++ b/sound/soc/codecs/ak4497.c
@@ -488,6 +488,8 @@ static int ak4497_hw_params(struct snd_pcm_substream *substream,
 	u8 dfs2;
 	int nfs1;
 	bool is_dsd = false;
+	int dsd_bclk;
+	u8 dsdsel0, dsdsel1;
 
 	if (pcm_format == SNDRV_PCM_FORMAT_DSD_U8 ||
 		pcm_format == SNDRV_PCM_FORMAT_DSD_U16_LE ||
@@ -505,6 +507,13 @@ static int ak4497_hw_params(struct snd_pcm_substream *substream,
 	dfs2 = snd_soc_read(codec, AK4497_05_CONTROL4);
 	dfs2 &= ~AK4497_DFS2;
 
+
+	dsdsel0 = snd_soc_read(codec, AK4497_06_DSD1);
+	dsdsel0 &= ~AK4497_DSDSEL0;
+
+	dsdsel1 = snd_soc_read(codec, AK4497_09_DSD2);
+	dsdsel1 &= ~AK4497_DSDSEL1;
+
 	if (!is_dsd) {
 		switch (nfs1) {
 		case 8000:
@@ -539,6 +548,33 @@ static int ak4497_hw_params(struct snd_pcm_substream *substream,
 		default:
 			return -EINVAL;
 		}
+	} else {
+		dsd_bclk = params_rate(params) *
+			params_physical_width(params);
+
+		switch (dsd_bclk) {
+		case 2822400:
+			dsdsel0 |= AK4497_DSDSEL0_2MHZ;
+			dsdsel1 |= AK4497_DSDSEL1_2MHZ;
+			break;
+		case 5644800:
+			dsdsel0 |= AK4497_DSDSEL0_5MHZ;
+			dsdsel1 |= AK4497_DSDSEL1_5MHZ;
+			break;
+		case 11289600:
+			dsdsel0 |= AK4497_DSDSEL0_11MHZ;
+			dsdsel1 |= AK4497_DSDSEL1_11MHZ;
+			break;
+		case 22579200:
+			dsdsel0 |= AK4497_DSDSEL0_22MHZ;
+			dsdsel1 |= AK4497_DSDSEL1_22MHZ;
+			break;
+		default:
+			return -EINVAL;
+		}
+
+		snd_soc_write(codec, AK4497_06_DSD1, dsdsel0);
+		snd_soc_write(codec, AK4497_09_DSD2, dsdsel1);
 	}
 
 	snd_soc_write(codec, AK4497_01_CONTROL2, dfs);
diff --git a/sound/soc/codecs/ak4497.h b/sound/soc/codecs/ak4497.h
index 99bd506..3ba6762 100644
--- a/sound/soc/codecs/ak4497.h
+++ b/sound/soc/codecs/ak4497.h
@@ -74,4 +74,17 @@
 #define AK4497_DFS2_48KHZ		(0x0 << 1)  // 30kHz to 216kHz
 #define AK4497_DFS2_384KHZ		(0x1 << 1)  // 384kHz, 768kHz to 108kHz
 
+
+#define AK4497_DSDSEL0			0x1
+#define AK4497_DSDSEL0_2MHZ		0x0
+#define AK4497_DSDSEL0_5MHZ		0x1
+#define AK4497_DSDSEL0_11MHZ		0x0
+#define AK4497_DSDSEL0_22MHZ		0x1
+
+#define AK4497_DSDSEL1			0x1
+#define AK4497_DSDSEL1_2MHZ		0x0
+#define AK4497_DSDSEL1_5MHZ		0x0
+#define AK4497_DSDSEL1_11MHZ		0x1
+#define AK4497_DSDSEL1_22MHZ		0x1
+
 #endif
-- 
1.7.9.5

